OneCommunityGlobal/HGNRest
# Angela - Twitter Pinterest Poster

# Description
This PR adds API endpoints for posting content on Pinterest and Twitter, enabling image extraction and posting capabilities (note: scheduling is not included in this PR). Currently, there are no related frontend PRs, as frontend development for these features is still in progress. Once the frontend is ready, these APIs will be integrated to complete the functionality.

## Main changes explained:
- Created `socialMediaController.js`: 
	•	Extract text and image URLs or base64 content from content.
	•	Post images to Pinterest using createPin.
	•	Post tweets with optional media content to Twitter using createTweet.
	•	Handle token retrieval for Pinterest and Twitter OAuth authentication.
- Created `socialMediaRouter.js`: Defines routes to handle Pinterest and Twitter API interactions, including posting images and obtaining access tokens.
- Updated `routes.js`: Integrated the new socialMediaRouter within the main app router configuration to expose the new API endpoints.

## How to test:
1. check into current branch
2. do `npm install` and `npm run dev` to run this PR locally
3. Access the endpoints:
- /api/getPinterestAccessToken - Test Pinterest token retrieval.
- /api/getTwitterAccessToken - Test Twitter token retrieval.
- /api/createPin - Test Pinterest posting functionality with sample data.
- /api/createTweet - Test Twitter posting functionality with sample data.

## Screenshots or videos of changes (Just temporary UI):
https://www.loom.com/share/e41dadeae7334dc9a20a1adff8b4b3fc?sid=ef77017e-d178-44b7-ba45-9d2105576708

## Note:
1. Pinterest Business Account:
- Ref: https://developers.pinterest.com/docs/getting-started/connect-app/
- For developing and testing, trial access is enough.
- For real world usage, you have to apply for standard access.

2. Twitter Developer Account Setup
- Apply for a Developer Account: Go to the [Twitter Developer Platform](https://developer.twitter.com/en/apply-for-access).
- Once approved, navigate to the [Developer Portal](https://developer.twitter.com/en/portal/dashboard).
- Create a new project and then an app within that project.
- Under “User authentication settings,” set the App permissions to “Read and Write.”
- Set the Callback URI to http://localhost:3000 or your application’s URL.

3. Set them as environment variables in your application (.env file):

Pinterest:
- REACT_APP_PINTEREST_CLIENT_ID
- REACT_APP_PINTEREST_CLIENT_SECRET

Twitter:
- REACT_APP_TWITTER_CLIENT_ID
- REACT_APP_TWITTER_CLIENT_SECRET
- REACT_APP_TWITTER_APP_KEY
- REACT_APP_TWITTER_APP_SECRET
- REACT_APP_TWITTER_ACCESS_TOKEN
- REACT_APP_TWITTER_ACCESS_SECRET
